libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2078,6 +2086,97 @@ void TypeChecker::typeCheckABIEncodeFunctions(
 	}
 }
 
+void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCall)
+{
+	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+
+	// Expecting first argument to be the function pointer and second to be a tuple.
+	if (arguments.size() != 2)
+	{
+		m_errorReporter.typeError(
+			6219_error,
+			_functionCall.location(),
+			"Expected two arguments: a function pointer followed by a tuple."
+		);
+		return;
+	}
+
+	auto const functionPointerType = dynamic_cast<FunctionTypePointer>(type(*arguments.front()));
+
+	if (!functionPointerType)
+	{
+		m_errorReporter.typeError(
+			5511_error,
+			arguments.front()->location(),
+			"Expected first argument to be a function pointer, not \"" +
+			type(*arguments.front())->canonicalName() +
+			"\"."
+		);
+		return;
+	}
+
+	if (functionPointerType->kind() != FunctionType::Kind::External)
+	{
+		string msg = "Function must be \"public\" or \"external\".";
+		SecondarySourceLocation ssl{};
+
+		if (functionPointerType->hasDeclaration())
+		{
+			ssl.append("Function is declared here:", functionPointerType->declaration().location());
+			if (functionPointerType->declaration().scope() == m_currentContract)
+				msg += " Did you forget to prefix \"this.\"?";
+		}
+
+		m_errorReporter.typeError(3509_error, arguments[0]->location(), ssl, msg);
+		return;
+	}
+
+	solAssert(!functionPointerType->takesArbitraryParameters(), "No checks possible for arbitrary parameters.");
+
+	// Tuples with only one component become that component
+	vector<ASTPointer<Expression const>> callArguments;
+
+	if (
+		auto const* argumentTuple = dynamic_cast<TupleExpression const*>(arguments[1].get());
+		argumentTuple && !argumentTuple->isInlineArray()
+	)
+		callArguments = decltype(callArguments){argumentTuple->components().begin(), argumentTuple->components().end()};
+	else
+		callArguments.push_back(arguments[1]);
+
+	if (functionPointerType->parameterTypes().size() != callArguments.size())
+		m_errorReporter.typeError(
+			7788_error,
+			_functionCall.location(),
+			"Expected " +
+			to_string(functionPointerType->parameterTypes().size()) +
+			" instead of " +
+			to_string(callArguments.size()) +
+			" components for the tuple parameter."
+		);
+
+	size_t const numParameters = min(callArguments.size(), functionPointerType->parameterTypes().size());
```

Comment:
You're only using `min()` to make the error non-fatal and be able to continue, right? I think I'd add a comment about that.